### PR TITLE
feat(ingest/oracle): support changing data dictionary (ALL_ or DBA_)

### PR DESCRIPTION
**Pull request by sleeperdeep**
actual discussion in [community](https://datahubspace.slack.com/archives/CUMUWQU66/p1652105092065869)
- change oracle plugin config: add additional field 'data_dictionary_mode' 
- change OracleInspectorObjectWrapper: add methods to extract data from dba_ views

Fixes https://github.com/datahub-project/datahub/issues/6711

## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: metadata-ingestion/src/datahub/ingestion/source/sql/oracle.py [hunk 1]
```diff
@@ -108,26 +144,163 @@ def __init__(self, inspector_instance: Inspector):
         # tables that we don't want to ingest into the DataHub
         self.exclude_tablespaces: Tuple[str, str] = ("SYSTEM", "SYSAUX")
 
+    def has_table(self, table_name, schema=None):
```
Comment (mayurinehate) [inline metadata-ingestion/src/datahub/ingestion/source/sql/oracle.py:147]: DataHub uses a few selective methods from inspector interface. You can look them up [here](https://github.com/datahub-project/datahub/blob/master/metadata-ingestion/src/datahub/ingestion/source/sql/sql_common.py) . Also listed below for quick reference. I believe it should suffice to implement only these, as required.

```
get_schema_names
get_table_names
get_view_names
get_pk_constraint
get_table_comment
get_view_definition
get_columns
get_foreign_keys
```

I doubt whether other methods implemented in wrapper here are needed.
These methods are not used by Datahub and therefore should be removed.
```
has_table
has_sequence
_resolve_synonym
_prepare_reflection_args
 get_temp_table_names
get_table_options
get_indexes
get_check_constraints
```

**Review by mayurinehate — CHANGES_REQUESTED** [review]
Thank you for the PR. I've left a few comments for changes.

Also, oracle tests are[ failing in CI](https://github.com/datahub-project/datahub/pull/8873/checks?check_run_id=17110894867). Can you please fix the tests ?

File: metadata-ingestion/src/datahub/ingestion/source/sql/oracle.py [hunk 2]
```diff
@@ -69,6 +91,11 @@ class OracleConfig(BasicSQLAlchemyConfig):
         default=False,
         description="Add oracle database name to urn, default urn is schema.table",
     )
+    # custom
+    data_dictionary_mode: Optional[str] = Field(
+        default='ALL',
+        description="The data dictionary views mode, to extract information about schema objects ('All' and 'DBA' views are supported). (https://docs.oracle.com/cd/E11882_01/nav/catalog_views.htm)"
```
[the remark this answers is not in the document]
Comment (sleeperdeep) [reply metadata-ingestion/src/datahub/ingestion/source/sql/oracle.py]: Fixed

File: metadata-ingestion/src/datahub/ingestion/source/sql/oracle.py [hunk 3]
```diff
@@ -97,6 +132,7 @@ def get_identifier(self, schema: str, table: str) -> str:
             return regular
 
 
+@inspector_wraper_usage_notificcation(class_usage_notification)
```
[the remark this answers is not in the document]
Comment (sleeperdeep) [reply metadata-ingestion/src/datahub/ingestion/source/sql/oracle.py]: Fixed

File: metadata-ingestion/src/datahub/ingestion/source/sql/oracle.py (hunk 1 above)
Comment (sleeperdeep) [reply metadata-ingestion/src/datahub/ingestion/source/sql/oracle.py]: Unused methods were deleted.

**Review by mayurinehate — CHANGES_REQUESTED** [review]
Hey @sleeperdeep , can you please fix oracle tests for your changes. It looks like the default behavior earlier was for `data_dictionay_mode=DBA`. I'm fine with current default as well but you may need to regenerate golden files for oracle tests. This should help there - https://datahubproject.io/docs/metadata-ingestion/developing#updating-golden-test-files

File: metadata-ingestion/src/datahub/ingestion/source/sql/sql_config.py [hunk 4]
```diff
@@ -90,6 +90,8 @@ class SQLCommonConfig(
     profiling: GEProfilingConfig = GEProfilingConfig()
     # Custom Stateful Ingestion settings
     stateful_ingestion: Optional[StatefulStaleMetadataRemovalConfig] = None
+    # Custom data_dictionary_mode
+    data_dictionary_mode: Optional[str] = None
```
Comment (mayurinehate) [inline metadata-ingestion/src/datahub/ingestion/source/sql/sql_config.py:94]: This should not be part of SQLCommonConfig. Can you remove this ?

File: metadata-ingestion/src/datahub/ingestion/source/sql/oracle.py [hunk 5]
```diff
@@ -299,7 +336,7 @@ def get_columns(
                 computed = None
 
             if identity_options is not None:
-                identity = self._inspector_instance.dialect._parse_identity_options(identity_options, default_on_nul)
+                identity = self.parse_identity_options(identity_options, default_on_nul)
```
Comment (mayurinehate) [inline metadata-ingestion/src/datahub/ingestion/source/sql/oracle.py:339]: Motivation behind this change is not clear . Can you help me understand why this change is needed ?

File: metadata-ingestion/src/datahub/ingestion/source/sql/oracle.py [hunk 6]
```diff
@@ -323,7 +360,7 @@ def get_columns(
         return columns
 
     def get_table_comment(
-            self, table_name: str, schema: str = None
+            self, table_name: Optional[str], schema: Optional[str] = None
```
Comment (mayurinehate) [inline metadata-ingestion/src/datahub/ingestion/source/sql/oracle.py:363]: same for this

**Comment by mayurinehate** [comment]
Hey @sleeperdeep did you get chance to go through the comments and update golden files ? Let me know if you have any queries. 

It would be great if you can verify oracle tests are passing locally by running them in development environment activated using [this](https://datahubproject.io/docs/metadata-ingestion/developing/#set-up-your-python-environment).
`pytest tests/unit/test_oracle_source.py`
`pytest tests/integration/oracle/test_oracle.py`

File: metadata-ingestion/src/datahub/ingestion/source/sql/sql_config.py (hunk 4 above)
Comment (sleeperdeep) [reply metadata-ingestion/src/datahub/ingestion/source/sql/sql_config.py]: @mayurinehate I`ve deleted it from SQLCommonConfig, but mypy returns attr-defined error.

**Comment by sleeperdeep** [comment]
Hi all! @mayurinehate, @hsheth2, @maggiehays, @mhw
Can you, please, review new commit?

**Comment by mayurinehate** [comment]
> Hi all! @mayurinehate, @hsheth2, @maggiehays, @mhw Can you, please, review new commit?

Thanks for the update @sleeperdeep . I can take a look early next week. Meanwhile - there seem to be [lint](https://github.com/datahub-project/datahub/actions/runs/7569093994/job/20756636917?pr=8873) and [test](https://github.com/datahub-project/datahub/actions/runs/7569093994/job/20756637129?pr=8873) failures, if you can take a look and fix those. 

If you have already setup and activated `venv` as described in developing [guide](https://datahubproject.io/docs/metadata-ingestion/developing/#set-up-your-python-environment), you should be able to run below command to find and fix lint issues

`black src/ tests/ && isort src/ tests/ && flake8 src/ tests/ && mypy src/ tests/` 

and below commands to find and fix test issues
`pytest tests/unit/test_oracle_source.py tests/integration/oracle/test_oracle.py`

Feel free to reach out to me on datahub slack if any questions.

File: metadata-ingestion/src/datahub/ingestion/source/sql/oracle.py [hunk 7]
```diff
@@ -52,13 +57,20 @@ def before_cursor_execute(conn, cursor, statement, parameters, context, executem
     cursor.outputtypehandler = output_type_handler
 
 
+def class_usage_notification(cls, func):
```
Comment (mayurinehate) [inline metadata-ingestion/src/datahub/ingestion/source/sql/oracle.py:60]: This can be removed, right ? Its not used anywhere.

File: metadata-ingestion/src/datahub/ingestion/source/sql/oracle.py [hunk 8]
```diff
@@ -184,8 +579,16 @@ def get_inspectors(self) -> Iterable[Inspector]:
...
             )
-            # To silent the mypy lint error
-            yield cast(Inspector, OracleInspectorObjectWrapper(inspector))
+            logger.info(
+                f'Data dictionary mode is: "{self.config.data_dictionary_mode}".'
+            )
+            # Sqlalchemy inspector uses ALL_* tables as per oracle dialect implementation.
+            # OracleInspectorObjectWrapper provides alternate implementation using DBA_* tables.
+            if self.config.data_dictionary_mode != "ALL":
+                yield cast(Inspector, OracleInspectorObjectWrapper(inspector))
```
Comment (mayurinehate) [inline metadata-ingestion/src/datahub/ingestion/source/sql/oracle.py:588]: Wondering, if we can rename `OracleInspectorObjectWrapper` to something like `OracleDBAInspectorObjectWrapper`, to make this more explicit in class name itself.

File: metadata-ingestion/src/datahub/ingestion/source/sql/oracle.py [hunk 9]
```diff
@@ -147,6 +169,377 @@ def get_table_names(
...
+            if computed is not None:
+                cdict["computed"] = computed
+            if identity is not None:
+                cdict["identity"] = identity
+
+            columns.append(cdict)
+        return columns
+
+    def get_table_comment(
+        self, table_name: Optional[str], schema: Optional[str] = None
```
Comment (mayurinehate) [inline metadata-ingestion/src/datahub/ingestion/source/sql/oracle.py:336]: `table_name` should not be optional, right ?

**Review by mayurinehate — APPROVED** [review]
Left few minor comments. Everything else looks good to me.
Thanks for the amazing work.

File: metadata-ingestion/src/datahub/ingestion/source/sql/oracle.py (hunk 9 above)
Comment (sleeperdeep) [reply metadata-ingestion/src/datahub/ingestion/source/sql/oracle.py:336]: @mayurinehate 
Yes, it should. But from perspective of logic of exctraction data from db. From perspective of structure of classes and it`s types [denormalize_name](https://github.com/sqlalchemy/sqlalchemy/blob/47716f5a45eb91361a5fdabb420144a1807ca8ae/lib/sqlalchemy/engine/interfaces.py#L1806) method in original sqlalchemy source and [denormalize_name](https://github.com/sqlalchemy/sqlalchemy2-stubs/blob/1ceec9ed1eb91e91800d35c5b8c42276b97f973b/sqlalchemy-stubs/engine/interfaces.pyi#L203) method in sqlalchemy-stup are differ. It leads to situation, where we chould choose what is better: use Optional[str] for main methods (get_schema_names, get_table_names, get_view_names, get_pk_constraint, get_table_comment, get_view_definition, get_columns, get_foreign_keys) or propagate "# type: ignore" throught all usages of denormalize_name method.

изображение

<->

изображение

File: metadata-ingestion/src/datahub/ingestion/source/sql/oracle.py (hunk 7 above)
Comment (sleeperdeep) [reply metadata-ingestion/src/datahub/ingestion/source/sql/oracle.py:60]: Yes, it can. Main idea behind this method consist in logging usage of wrapper class.

File: metadata-ingestion/src/datahub/ingestion/source/sql/oracle.py (hunk 8 above)
Comment (sleeperdeep) [reply metadata-ingestion/src/datahub/ingestion/source/sql/oracle.py:588]: Oracle allows three levels of access (based on levels of privilege): ALL, DBA, USER. Right now ALL mode is using as default. DBA mode is implemented by this PR. But pottentially, mode USER can be implemented as future contribution. In this case renaming of this class will not give nothing. I can extend docstring, if it is necessary.

Comment (mayurinehate) [reply metadata-ingestion/src/datahub/ingestion/source/sql/oracle.py:580]: I see.

File: metadata-ingestion/src/datahub/ingestion/source/sql/oracle.py (hunk 9 above)
Comment (mayurinehate) [reply metadata-ingestion/src/datahub/ingestion/source/sql/oracle.py:328]: I don't know if one is better than other. @hsheth2  - any thoughts ? Also - could you please take a look and get this PR merged ? Its looks all good to me.

Comment (hsheth2) [reply metadata-ingestion/src/datahub/ingestion/source/sql/oracle.py:328]: For all the main methods (e.g. get_table_names, get_view_names, get_pk_constraint, get_table_comment, get_view_definition, get_columns, get_foreign_keys) - these do actually require `table_name: str`, so declaring it to be optional feels misleading. I'd prefer something like below, or using `type: ignore` if the below doesn't work for some reason

```python
def get_table_comment(self, table_name: str, schema: Optional[str] = None)
  denormalized_table_name: Optional[str] = self._inspector_instance.dialect.denormalize_name(table_name)
  assert denormalized_table_name is not None
  denormalized_table_name = table_name  # or maybe just use `denormalized_table_name` in the remainder of the method
```

File: metadata-ingestion/src/datahub/ingestion/source/sql/oracle.py [hunk 10]
```diff
@@ -147,6 +161,377 @@ def get_table_names(
...
+            schema or self.default_schema_name
+        )
+
+        if schema is None:
+            schema = self._inspector_instance.dialect.default_schema_name
+
+        COMMENT_SQL = """
+            SELECT comments
+            FROM dba_tab_comments
+            WHERE table_name = CAST(:table_name AS VARCHAR(128))
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/sql/oracle.py:341]: eventually we'll want to replace this with a single bulk fetch for the full schema, instead of fetching each table one at a time

**Review by hsheth2 — COMMENTED** [review]
mostly looks good, the main thing is tweaking those `Optional[str]` annotations

File: metadata-ingestion/src/datahub/ingestion/source/sql/oracle.py (hunk 9 above)
Comment (sleeperdeep) [reply metadata-ingestion/src/datahub/ingestion/source/sql/oracle.py]: @hsheth2 
I fixed it. Agree, this approach is better.

File: metadata-ingestion/src/datahub/ingestion/source/sql/oracle.py [hunk 11]
```diff
@@ -147,6 +161,377 @@ def get_table_names(
...
+        params = {"view_name": view_name}
+        text = "SELECT text FROM dba_views WHERE view_name=:view_name"
+
+        if schema is not None:
+            text += " AND owner = :schema"
+            params["schema"] = schema
+
+        rp = self._inspector_instance.bind.execute(sql.text(text), params).scalar()
+        if rp:
+            if py2k:
```
[the remark this answers is not in the document]
Comment (sleeperdeep) [reply metadata-ingestion/src/datahub/ingestion/source/sql/oracle.py]: Agree. Was removed from module.

File: metadata-ingestion/src/datahub/ingestion/source/sql/oracle.py [hunk 12]
```diff
@@ -147,6 +161,377 @@ def get_table_names(
...
+            AND owner = CAST(:schema_name AS VARCHAR(128))
+        """
+
+        c = self._inspector_instance.bind.execute(
+            sql.text(COMMENT_SQL), dict(table_name=table_name, schema_name=schema)
+        )
+
+        return {"text": c.scalar()}
+
+    def _get_constraint_data(
```
[the remark this answers is not in the document]
Comment (sleeperdeep) [reply metadata-ingestion/src/datahub/ingestion/source/sql/oracle.py:358]: Yes. Main idea consist in accessing other schema (DBA instead of ALL) without global changes of original modules.

File: metadata-ingestion/src/datahub/ingestion/source/sql/oracle.py (hunk 10 above)
Comment (sleeperdeep) [reply metadata-ingestion/src/datahub/ingestion/source/sql/oracle.py:347]: I agree with you. This approach (single query to fetch all info in one go) is more preferable. Idea of this PR was different. I can test it and contribute in next PR.

**Review by hsheth2 — APPROVED** [review]

**Comment by hsheth2** [comment]
The `test_create_list_get_ingestion_execution_request` is a known issue, so merging this in.

Thanks @sleeperdeep for incredible contribution 🎉
